### PR TITLE
Construct FOLIO Inventory Record and adds to FOLIO

### LINKED ISSUE #2: Add new Inventory Record in FOLIO
Use Okapi Inventory API to create FOLIO Records (can be [Instance](https://wiki.folio.org/pages/viewpage.action?pageId=54888353)

This will likely need an Airflow [Connection](https://airflow.apache.org/docs/apache-airflow/stable/howto/connection.html) for each FOLIO's environment Okapi endpoint that has `mod-inventory` enabled for use.

Currently have an open ticket in EBSCO https://github.com/FOLIO-FSE/MARC21-To-FOLIO/issues/77 to see if they will publish their module to pypi so we can leverage some of their existing Python classes without needing a local copy of their repo.

### PR DESCRIPTION
**Pull request by jermnelson**
FIXES #2 

TODO:
- [x] Finish transforms
- [x] Finish tests
- [x] New FOLIO Record

### CODE REVIEW

File: ils_middleware/tasks/folio/new.py [hunk 1]
```diff
@@ -0,0 +1,86 @@
...
+        "errors"
+    )[0]["message"].startswith("id value already exists"):
+        _overlay_folio_records(
+            okapi_instance_url,
+            headers=headers,
+            payload=payload,
+        )
+        _push_to_xcom(records, task_instance)
+    else:
+        logger.error(f"New records failed errors: {new_record_result.json()}")
```
Comment (jcoyne) [inline ils_middleware/tasks/folio/new.py:63]: Should we have tests for this case?

File: ils_middleware/tasks/folio/new.py [hunk 2]
```diff
@@ -0,0 +1,86 @@
...
+
+def post_folio_records(**kwargs):
+    """Creates new records in FOLIO"""
+    task_instance = kwargs["task_instance"]
+    jwt = kwargs["token"]
+    task_groups = ".".join(kwargs["task_groups_ids"])
+
+    task_id = "build-folio"
+    if len(task_groups) > 0:
+        task_id = f"{task_groups}.{task_id}"
```
Comment (jcoyne) [inline ils_middleware/tasks/folio/new.py:75]: It's showing this isn't covered by a test, is that right?

File: ils_middleware/tasks/folio/new.py [hunk 3]
```diff
@@ -0,0 +1,86 @@
+import logging
```
Comment (jcoyne) [inline ils_middleware/tasks/folio/new.py:1]: Please add some documentation about what this does.

File: pyproject.toml [hunk 4]
```diff
@@ -16,6 +16,8 @@ hvac = "^0.11.2"
 rdflib = "^6.0.2"
 honeybadger = "^0.7.1"
 types-requests = "^2.26.1"
+folio-uuid = "^0.2.0"
```
Comment (jcoyne) [inline pyproject.toml:19]: What is the purpose of this.  Why is it called a "uuid", but it doesn't look like a UUID?

File: ils_middleware/tasks/folio/mappings/bf_instance.py [hunk 5]
```diff
@@ -1,16 +1,3 @@
...
-SELECT ?date
-WHERE {{
-    <{bf_instance}> a bf:Instance .
-    <{bf_instance}> bf:provisionActivity ?activity .
-    ?activity a bf:Publication .
-    ?activity bf:date ?date .
-}}
-"""
-
 identifier = """PREFIX bf: <http://id.loc.gov/ontologies/bibframe/>
```
Comment (jcoyne) [inline ils_middleware/tasks/folio/mappings/bf_instance.py:1]: I would appreciate some documentation in this file about it's purpose.

File: ils_middleware/tasks/folio/map.py [hunk 6]
```diff
@@ -14,17 +14,9 @@
         "uri": "work",
         "class": "bf:Person",
     },
-    "dateOfPublication": {
```
Comment (jcoyne) [inline ils_middleware/tasks/folio/map.py:17]: Is this something that FOLIO doesn't support?  Is this mapping complete or do we think there is more work here?

File: ils_middleware/tasks/folio/login.py [hunk 7]
```diff
@@ -1,20 +1,23 @@
...
 
+logger = logging.getLogger(__name__)
+
 
 def FolioLogin(**kwargs) -> SimpleHttpOperator:
     """Logs into FOLIO and returns Okapi token."""
     url = kwargs["url"]
     username = kwargs["username"]
     password = kwargs["password"]
+    tenant = kwargs["tenant"]
```
Comment (jcoyne) [inline ils_middleware/tasks/folio/login.py:15]: 👍

File: ils_middleware/tasks/folio/build.py [hunk 8]
```diff
@@ -0,0 +1,312 @@
...
+    for row in values:
+        subjects.append(row[0])
+
+    return "subjects", subjects
+
+
+def _title(**kwargs) -> tuple:
+    values = kwargs.get("values")
+    if isinstance(values, str):
+        values = [
```
Comment (jcoyne) [inline ils_middleware/tasks/folio/build.py:200]: Do we need a test for this case?

File: ils_middleware/tasks/folio/build.py [hunk 9]
```diff
@@ -0,0 +1,312 @@
+import datetime
```
Comment (jcoyne) [inline ils_middleware/tasks/folio/build.py:1]: Can you add documentation about what this file is for?

File: ils_middleware/dags/stanford.py [hunk 10]
```diff
@@ -209,7 +239,8 @@ def dag_failure_callback(ctx_dict) -> None:
                     "SIRSI": [
                         "process_symphony.post_new_symphony",
                         "process_symphony.post_overlay_symphony",
-                    ]
+                    ],
+                    "FOLIO": ["process_folio.new-folio-records"],
```
Comment (jcoyne) [inline ils_middleware/dags/stanford.py:243]: Do we not also need an overlay? If folio can handle both cases in one handler, should this have a better name?

**Review by jcoyne — CHANGES_REQUESTED** [review]

File: ils_middleware/tasks/folio/new.py [hunk 11]
```diff
@@ -0,0 +1,86 @@
...
+    resources = task_instance.xcom_pull(key="resources", task_ids="sqs-message-parse")
+
+    inventory_records = []
+    for instance_uri in resources:
+        inventory_records.append(
+            task_instance.xcom_pull(key=instance_uri, task_ids=task_id)
+        )
+
+    _post_to_okapi(records=inventory_records, jwt=jwt, **kwargs)
+    # Post back to
```
Comment (aaron-collier) [inline ils_middleware/tasks/folio/new.py:86]: cruft? or should this be a more explanatory comment?

File: ils_middleware/tasks/folio/build.py (hunk 9 above)
Comment (jermnelson) [reply ils_middleware/tasks/folio/build.py:4]: Done, added module documentation

File: ils_middleware/tasks/folio/map.py (hunk 6 above)
Comment (jermnelson) [reply ils_middleware/tasks/folio/map.py:17]: The `dateOfPublication` has been absorbed into the `publication` and is now in the FOLIO expected format for a publication.

File: ils_middleware/tasks/folio/mappings/bf_instance.py (hunk 5 above)
Comment (jermnelson) [reply ils_middleware/tasks/folio/mappings/bf_instance.py:5]: Added module descriptions to both `folio.mappings.bf_instance` and `folio.mappings.bf_work`.

File: ils_middleware/tasks/folio/new.py (hunk 1 above)
Comment (jermnelson) [reply ils_middleware/tasks/folio/new.py:46]: Done,  we now test for this case in `tests.test_new_inventory.test_raised_error`.

File: ils_middleware/tasks/folio/new.py (hunk 2 above)
Comment (jermnelson) [reply ils_middleware/tasks/folio/new.py:58]: Now covered.

File: pyproject.toml (hunk 4 above)
Comment (jermnelson) [reply pyproject.toml:19]: This generates module from Ebsco generates a deterministic uuid for FOLIO records based on the okapi_uri, a enumeration value based on the record type, and an ILS id, in our case the Sinopia BF Instance URL. 
```python
> from folio_uuid import FOLIONamespaces, FolioUUID

> okapi_uri = "https://okapi-folio.dev.edu/"
> instance_uri = "https://api.development.sinopia.io/resource/0000-1111-2222-3333"

> print(FolioUUID(okapi_url, FOLIONamespaces.instances, instance_uri))
98a0337a-ec22-53aa-8ffc-933a86d10159
```

File: ils_middleware/tasks/folio/new.py (hunk 3 above)
Comment (jermnelson) [reply ils_middleware/tasks/folio/new.py:6]: Done.

File: ils_middleware/dags/stanford.py (hunk 10 above)
Comment (jermnelson) [reply ils_middleware/dags/stanford.py]: Done, with @jgreben confirming that the post to the endpoint supports `upsert`, renamed this task.

File: ils_middleware/tasks/folio/build.py (hunk 8 above)
Comment (jermnelson) [reply ils_middleware/tasks/folio/build.py]: Kruft from early testing; removed

File: ils_middleware/tasks/folio/new.py (hunk 11 above)
Comment (jermnelson) [reply ils_middleware/tasks/folio/new.py]: Removed!

File: ils_middleware/tasks/folio/map.py (hunk 6 above)
Comment (jcoyne) [reply ils_middleware/tasks/folio/map.py:17]: 👍

**Review by jcoyne — APPROVED** [review]

**Pull request merged**